src/graph_ir.rs

```diff
@@ -145,17 +146,17 @@ impl GraphIR {
             return Err(CompileGraphError::CycleDetected);
         }
 
-        let mut stack = self.roots().cloned().collect::<Vec<_>>();
+        let mut stack = self.roots().cloned().collect::<VecDeque<_>>();
         let mut visited = FnvHashSet::default();
         visited.reserve(self.nodes.len());
 
         self.schedule.clear();
 
-        while let Some(node) = stack.pop() {
+        while let Some(node) = stack.pop_front() {
             if !visited.contains(&node.id) {
                 visited.insert(node.id);
                 for next in self.outgoing(&node) {
-                    stack.push(next.clone());
+                    stack.push_back(next.clone());
```

Comment:
The alternative fix is to only add to the schedule and mark permanently if all the children have been visited, the bug is that if this were a recursive call storing state on the call stack it would probably work. Using a queue instead will work though, I believe it converts this from DFS to BFS.

Reply:
Ok, so do you want this changed, or keep it as a queue for now?